scipy/optimize/zeros.c

```diff
@@ -6,6 +6,48 @@
 
 #include "Python.h"
 #include <setjmp.h>
+#include <numpy/npy_math.h>
+
+//#include <stdio.h>
+#if PY_VERSION_HEX >= 0x03000000
+    #define PyString_FromString PyBytes_FromString
+    #define PyString_Concat PyBytes_Concat
+    #define PyString_AsString PyBytes_AsString
+    #define PyInt_FromLong PyLong_FromLong
+    #define PyInt_AsLong PyLong_AsLong
+#endif
+
+#ifndef FALSE
+#define FALSE 0
+#endif /* FALSE */
+#ifndef TRUE
+#define TRUE 1
+#endif /* TRUE */
+
+#include "ccallback.h"
+#include "Zeros/zeros.h"
+
+/*
+ * Caller entry point functions
+ */
+
+typedef enum {
+    CB_D = 1,         // f(x)
+    CB_D_D = 2,       // f(x, double)
+    CB_D_I_ND = 10,   // f(x, int n, double *)   I.e. pointer to n doubles
+    CB_D_VOIDSTAR = 100,     // f(x, void *) or f(x, const void *)
+    CB_D_USERDATA = 1001     // f(x, void *) or f(x, const void *) using UserData
+} zeros_signature_t;
+
+static ccallback_signature_t signatures[] = {
+    {"double (double)", CB_D},
+    {"double (double, double)", CB_D_D},
+    {"double (double, int, double *)", CB_D_I_ND},
```

Comment:
Should also add:
```
#if NPY_SIZEOF_SHORT == NPY_SIZEOF_INT
    {"double (double, short, double *)", CB_D_I_ND},
#endif
#if NPY_SIZEOF_LONG == NPY_SIZEOF_INT
    {"double (double, long, double *)", CB_D_I_ND},
#endif
```
It runs on manual code generation, for simplicity...

Reply:
Thanks for the comments.  I definitely consider this PR to be WIP for a few reasons, and the comments are helpful.  Here are some issues.

1. `user_data`, `args=` or both?
2. `toms748` is written in Python, not Cython and doesn't take a LowLevelCallable. Not good from the user experience.
3. The `newton` solver raises 3 issues: It's written in Python. It accepts complex-valued `x` and `f(x)`.  And it should able return 2 values --- both `f(x)` and `f'(x)` --- in a single call, just as `zeros.root` allows, and gh-8890 for `zeros.root_scalar`.

Expanding on these:
1. The user appears to have two ways to specify additional arguments to a parametrized function. One is by using the `args=` approach, and having those be interpreted/converted. The other is to use the `user_data` in the LowLevelCallable.  I started down the latter direction but then leaned towards the former as `user_data` seems to require a separate LowLevelCallable instance for each specialization, which appeared to make the approach involve more cumbersome for the user, especially when solving multiple equations in parallel. 

2. How does one call a LowLevelCallable from pure Python?  `test_call_cython` and `test_thunk_cython` seem to do it from Cython in `_lib/_ccallback_c.pyx`.  Is that the only way, to introduce a Cython thunk function?

3. The signatures for `newton` would be multiple, even without consideration of user_data.  E.g.
- double f(double)
- complex f(complex)
- complex  f(complex, *complex) // to return the 1st derivative or possibly
- complex[2] f(complex) // alternative way to return the 1st derivative